File: app/src/main/java/com/krev/trycrypt/adapters/PlacesAdapter.java
package com.krev.trycrypt.adapters;

import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.BitmapDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.krev.trycrypt.R;
import com.krev.trycrypt.server.model.entity.Place;
import com.krev.trycrypt.utils.Consumer;
import com.krev.trycrypt.utils.DownloadImageTask;

import java.util.Arrays;
import java.util.List;

/**
 * Created by Dima on 14.11.2016.
 */

public class PlacesAdapter extends BaseAdapter {
    private final LayoutInflater layoutInflater;
    private List<Place> places;
    private List<ImageView> imageViews;

    public PlacesAdapter(List<Place> places, Activity activity) {
        this.places = places;
        this.imageViews = Arrays.asList(new ImageView[this.places.size()]);
        this.layoutInflater = (LayoutInflater) activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    @Override
    public int getCount() {
        return places.size();
    }

    @Override
    public Object getItem(int i) {
        return places.get(i);
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public View getView(int position, View contentView, ViewGroup viewGroup) {
        ViewHolder holder = new ViewHolder();
        if (contentView == null) {
            contentView = layoutInflater.inflate(R.layout.place, viewGroup, false);
        }
        holder.icon = (ImageView) contentView.findViewById(R.id.place_icon);
        holder.name = (TextView) contentView.findViewById(R.id.place_name);
        Place place = places.get(position);
        if (imageViews.get(position) == null) {
            imageViews.set(position, holder.icon);
            try {
                new DownloadImageTask(imageViews.get(position),
                        new Consumer<ImageView>() {
                            @Override
                            public void accept(ImageView o) {
                                notifyDataSetChanged();
                            }
                        }).execute(place.getPhoto());
            } catch (NullPointerException ignored) {
            }
        } else {
            holder.icon.setImageBitmap(((BitmapDrawable) imageViews.get(position).getDrawable()).getBitmap());
        }
        holder.name.setText(place.getName());
        contentView.setTag(holder);
        return contentView;
    }

    private class ViewHolder {
        ImageView icon;
        TextView name;
    }
}

File: app/src/main/java/com/krev/trycrypt/adapters/PlacesAdapter.kt
package com.krev.trycrypt.adapters

import android.content.Intent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.krev.trycrypt.R
import com.krev.trycrypt.activity.MapActivity
import com.krev.trycrypt.application.Config
import com.krev.trycrypt.server.model.entity.Place
import com.krev.trycrypt.utils.PhotoUtils
import com.mapbox.mapboxsdk.camera.CameraPosition

class PlacesAdapter : CustomAdapter<Place>() {
    override var layout: Int = R.layout.listview_place

    override fun photo(item: Place, view: ImageView) = PhotoUtils.getAwait(item.photo, item.id, view)

    override fun View.holder() = this.apply {
        tag = ViewHolder(
                (findViewById(R.id.place_icon) as ImageView),
                (findViewById(R.id.place_name) as TextView),
                (findViewById(R.id.place_creator_icon) as ImageView),
                (findViewById(R.id.place_creator_name) as TextView))
    }

    override fun View.data(item: Place) = this.apply {
        setOnClickListener {
            Config.camera = CameraPosition.Builder().target(MapActivity.convert(item.location)).zoom(16.0).build()
            Config.activity!!.startActivity(Intent(Config.activity!!, MapActivity::class.java))
            Config.activity!!.finish()
        }
        (tag as ViewHolder).apply {
            photo(item, icon)
            name.text = item.name
            Config.findUser(item.owner).apply { PhotoUtils.getAwait(item.photo, item.id, creatorIcon) }
            creatorName.text = Config.findUser(item.owner).name
        }
    }

    data class ViewHolder(val icon: ImageView,
                          val name: TextView,
                          val creatorIcon: ImageView,
                          val creatorName: TextView)
}